src/composables/useSocksProxy.ts

```diff
@@ -1,99 +1,151 @@
-import { computed } from 'vue';
-import { proxy } from 'webextension-polyfill';
-import { sendMessage } from 'webext-bridge/popup';
+import { Ref, computed } from 'vue';
 
-import useStore from './useStore';
+import {
+  ProxyDetails,
+  ProxyInfo,
+  ProxyInfoType,
+  ProxyOperationArgs,
+} from '@/helpers/socksProxy.types';
 
+import useActiveTab from '@/composables/useActiveTab';
 import useConnection from '@/composables/useConnection';
-import getSocksIpForProtocol from '@/composables/utils/getSocksIpForProtocol';
+import useStore from '@/composables/useStore';
 
-const { connection, updateConnection } = useConnection();
-
-const DEFAULT_PORT = 1080;
-
-const baseConfig = {
-  // FIXME: Allow disabling Proxy DNS in the settings
+const baseConfig: Partial<ProxyInfo> = {
+  port: 1080,
   proxyDNS: true,
-  proxyType: 'manual',
 };
 
-const { socksDetails } = useStore();
+const socksIp = '10.64.0.1';
 
-const enableProxy = async () => {
-  const socksIp = getSocksIpForProtocol(connection.value.protocol);
+const { activeTabHost } = useActiveTab();
+const { updateConnection } = useConnection();
+const { globalProxy, globalProxyDetails, hostProxies, hostProxiesDetails } = useStore();
 
-  await proxy.settings.set({
-    value: {
-      ...baseConfig,
-      socks: `${socksIp}:${DEFAULT_PORT}`,
-    },
-  });
+const currentHostProxyDetails = computed(
+  () => hostProxiesDetails.value[activeTabHost.value] || null,
+);
 
-  await updateConnection();
+const globalProxyEnabled = computed(() => globalProxyDetails.value.socksEnabled);
+const currentHostProxyEnabled = computed(
+  () => currentHostProxyDetails.value?.socksEnabled ?? false,
+);
 
-  socksDetails.value = {
-    socksEnabled: true,
-    protocol: connection.value.protocol,
-    server: connection.value.server,
-    proxyDNS: baseConfig.proxyDNS,
+const globalProxyDNSEnabled = computed(() => globalProxy.value?.proxyDNS ?? false);
+const currentHostProxyDNSEnabled = computed(() => currentHostProxyDetails.value?.proxyDNS ?? false);
+
+const setProxyDetails = (proxyDetails: Ref<ProxyDetails>, enabled: boolean) => {
+  proxyDetails.value = {
+    ...proxyDetails.value,
+    socksEnabled: enabled,
   };
+};
 
-  sendMessage('update-socks', {}, 'background');
+const toggleGlobalProxy = () => {
+  setProxyDetails(globalProxyDetails, !globalProxyDetails.value.socksEnabled);
+};
+const toggleCurrentHostProxy = () => {
+  hostProxiesDetails.value[activeTabHost.value] = {
```

Comment:
Can't you just do `hostProxiesDetails.value[activeTabHost.value].socksEnabled = !currentHostProxyEnabled.value` instead of spreading the current value?